Add ballot calculating code

lisa-sa-li opened the pull request:
## Overview

- Resolves #7 
- Original ballot election script in `ballot.js`
- Direct JS -> Python translation in `ballot.py`
- Revised Python ballot election script in `ballot_new.py`


## Unit Tests Created

Tests written for 3 cases
- Case 1: YES/NO election (simplest case) i.e. a referendum or one person running (written by @albogdan)
- Case 2: Single seat election (written by @arminale)
- Case 3: Multi-seat election with more than two candidates (written by @lisa-sa-li)

Reached 100% coverage in `ballot.py`


## Steps to QA

- Run unit tests

lisa-sa-li committed:
add ballot calculating code

lisa-sa-li committed:
add ballot calculating code

albogdan committed:
Updated in pip and ran tests (#128)

albogdan committed:
- Added map_candidates_in_ballots_to_choices function. Temporary before moving. Have to create admin view first
- Black formatting

albogdan committed:
- Added map_candidates_in_ballots_to_choices function. Temporary before moving. Have to create admin view first
- Black formatting

albogdan committed:
- Created 3 unit tests for the first if-statement. These can be used as templates for the elif and else statement unit tests
- Renamed the file to 'test_ballot' so that django test framework detects it automatically when running all unit tests

albogdan committed:
- Created 3 unit tests for the first if-statement. These can be used as templates for the elif and else statement unit tests
- Renamed the file to 'test_ballot' so that django test framework detects it automatically when running all unit tests

albogdan committed:
- Added a serializer to convert from the representation of ballots in the database to the one that the calculate_results function takes
- Note that the input must be a queryset list

albogdan committed:
- Added a serializer to convert from the representation of ballots in the database to the one that the calculate_results function takes
- Note that the input must be a queryset list

albogdan committed:
- Moved the map_candidates_in_ballots_to_choices from the ballot.py file to the serializers.py file and the BallotSerializer class as a staticmethod
	- Updated calls to it from test_ballot.py

albogdan committed:
- Moved the map_candidates_in_ballots_to_choices from the ballot.py file to the serializers.py file and the BallotSerializer class as a staticmethod
	- Updated calls to it from test_ballot.py

albogdan committed:
Black formatting for ballot_new.py

albogdan committed:
Black formatting for ballot_new.py

albogdan committed:
Added a function to create a random number of voters

albogdan committed:
Added a function to create a random number of voters

lisa-sa-li commented on skule_vote/backend/test_ballot.py:
```diff
@@ -0,0 +1,312 @@
+        self._generate_random_voters(num_voters=NUM_VOTERS)
+        voters = Voter.objects.all()
+        ballots = [
+            {  # First vote (2 candidates ranked)
+                "voter": voters[0],
+                "candidate": candidate,
+                "rank": 0,
+                "election": referendum,
+            },
+            {
+                "voter": voters[0],
+                "candidate": ron,
+                "rank": 1,
+                "election": referendum,
+            },
```
you can only vote for 1 option, Yes (for the candidate/referenda) or No (ron). So the len(ranking) will only be 1 or 0

arminale commented on skule_vote/backend/serializers.py:
```diff
@@ -1,7 +1,43 @@
 from rest_framework import serializers
 
 
+class BallotSerializer(serializers.BaseSerializer):
```
This is outdated code :( 
It is missing the entire ballot submission endpoint and infra which includes a different ballot serializer. Need to pull from develop and rename this to BallotResultSerializer or something

arminale commented on skule_vote/backend/test_ballot.py:
```diff
@@ -0,0 +1,312 @@
+        self.assertEqual(results["totalVotes"], len(voters) - NUM_SPOILED)
+
+    # YES/NO election with some ballots being errors
+    def test_one_candidate_ballot_errored(self):
```
I wonder if this should error out instead of returning the error count? probably something to discuss with them?

albogdan committed:
Merged in BallotSerializer from develop
Renamed the new BallotSerializer being used for results calculations to BallotResultsCalculationSerializer
Reordered classes to be alphabetical

albogdan committed:
Removed generate_random_voters function I wrote because it was already written in the rebase from develop

albogdan committed:
Fixed tests to work with the new rebased changes

albogdan committed:
Fixed the dictionary key to be the same as the one for ballot.py, for easy testing
	Now just have to modify the import statement in test_ballot.py in order to test with any file

albogdan committed:
In a yes-no election, each voter only votes EITHER yes or no or spoils their ballot. There is no ranking yes or no. Updated test cases to reflect this.

albogdan committed:
Merge branch '7-generateElectionResults' of github.com:albogdan/skule-vote into 7-generateElectionResults

albogdan committed:
Fixed a small issue I forgot about when merging

lisa-sa-li commented on skule_vote/backend/test_ballot.py:
```diff
@@ -0,0 +1,376 @@
+        self._generate_voters(count=NUM_VOTERS)
+        voters = Voter.objects.all()
+        ballots = [
+            {  # First vote (2 candidates ranked)
```
nit: gotta update this comment

lisa-sa-li commented on skule_vote/backend/test_ballot.py:
```diff
@@ -0,0 +1,376 @@
+        self._generate_voters(count=NUM_VOTERS)
+        voters = Voter.objects.all()
+        ballots = [
+            {  # First vote (2 candidates ranked)
```
nit: same here

lisa-sa-li replied:
actually since they all rank 1 candidate, i think the comment `First vote` is sufficient

lisa-sa-li committed:
add case 3 starter code for unit test

lisa-sa-li committed:
add test cases for case 3

lisa-sa-li committed:
add 2 new cases for Case 3 and make comments clearer

lisa-sa-li committed:
clean up creating candidates, ballots, and results with helper functions

lisa-sa-li committed:
update assertEqual with candidate.name instead of "Alex"

arminale committed:
Added tests for 1 seat elections

lisa-sa-li committed:
add new case 3 tests to increase coverage

lisa-sa-li committed:
Merge branch '7-generateElectionResults' of github.com:albogdan/skule-vote into 7-generateElectionResults

lisa-sa-li committed:
push broken test

lisa-sa-li commented on skule_vote/backend/ballot.py:
```diff
@@ -0,0 +1,493 @@
+                                print(
+                                    f"ERROR - Ballot contained invalid ranking: {ranking[currentRanking]}"
+                                )
+                                # Figure out what we're doing in this edge case, below is temp for now
+                                keepChecking = False
```
WE NEED TO DECIDE ON WHAT TO DO WHEN THERE'S AN INVALID BALLOT????

lisa-sa-li committed:
push 2 tests in case 3 with weird behavior, explained

lisa-sa-li committed:
new test for increase coverage of backwards elimination process

arminale committed:
Added test for malformed ballots in single seat elections

lisa-sa-li committed:
add 2 more tests in case 3 to increase backward elimination process coverage

lisa-sa-li committed:
clean up case 1 & 3 tests with helper functions

albogdan committed:
Fixed issue where RON was sometimes not the first item in the list, leading to randomly failed tests

albogdan committed:
Black formatting

lisa-sa-li commented on skule_vote/backend/test_ballot.py:
```diff
@@ -0,0 +1,1637 @@
+
+        NUM_VOTERS = 12
+        NUM_SPOILED = 0
+        NUM_ERRORED = 1
```
@arminale, this constant isn't used in the test. Can you change the assertEqual to: `self.assertEqual(results["rounds"][0][candidate2.name], 2 - NUM_ERRORED)` just so we can easily see which candidate lost the vote bc of the ballot? (It might be candidate1, i'm not 100% sure cause I didn't throughly check the testcase)

lisa-sa-li committed:
clean up comments in test_ballot.py

lisa-sa-li committed:
clean up comments and code in ballot_new

arminale committed:
improved readability by using num_error constant instead of hardcoding

arminale committed:
fixed stupid typo and updated test to match
